microphone/microphone_driver.c

```diff
@@ -0,0 +1,815 @@
+{
+   settings_t *settings   = (settings_t*)settings_data;
+   microphone_driver_state_t *mic_st = &mic_driver_st;
+   float slowmotion_ratio = settings->floats.slowmotion_ratio;
+   bool verbosity_enabled = verbosity_is_enabled();
+   size_t max_frames   = AUDIO_CHUNK_SIZE_NONBLOCKING * AUDIO_MAX_RATIO * slowmotion_ratio;
+
+   if (!settings->bools.microphone_enable)
+   { /* If the user has mic support turned off... */
+      mic_st->flags &= ~MICROPHONE_DRIVER_FLAG_ACTIVE;
+      RARCH_WARN("[Microphone]: Refused to initialize microphone driver because it's disabled in the settings\n");
+      return false;
+   }
+
+   convert_s16_to_float_init_simd();
+   convert_float_to_s16_init_simd();
+
+   if (!(microphone_driver_find_driver(settings,
+                                       "microphone driver", verbosity_enabled)))
+   {
+      RARCH_ERR("[Microphone]: Failed to initialize microphone driver. Will continue without mic input.\n");
+      goto error;
+   }
+
+   mic_st->input_frames_length = max_frames * sizeof(float);
+   mic_st->input_frames = (float*)memalign_alloc(64, mic_st->input_frames_length);
+   if (!mic_st->input_frames)
+      goto error;
+
+   mic_st->converted_input_frames_length = max_frames * sizeof(float);
+   mic_st->converted_input_frames = (float*)memalign_alloc(64, mic_st->converted_input_frames_length);
+   if (!mic_st->converted_input_frames)
+      goto error;
+
+   /* Need room for dual-mono frames */
+   mic_st->dual_mono_frames_length = max_frames * sizeof(float) * 2;
+   mic_st->dual_mono_frames = (float*)memalign_alloc(64, mic_st->dual_mono_frames_length);
+   if (!mic_st->dual_mono_frames)
+      goto error;
+
+   mic_st->resampled_frames_length = max_frames * sizeof(float) * 2;
+   mic_st->resampled_frames = (float*) memalign_alloc(64, mic_st->resampled_frames_length);
+   if (!mic_st->resampled_frames)
+      goto error;
+
+   mic_st->resampled_mono_frames_length = max_frames * sizeof(float);
+   mic_st->resampled_mono_frames = (float*) memalign_alloc(64, mic_st->resampled_mono_frames_length);
+   if (!mic_st->resampled_mono_frames)
+      goto error;
+
+   mic_st->final_frames_length = max_frames * sizeof(int16_t);
+   mic_st->final_frames = (int16_t*) memalign_alloc(64, mic_st->final_frames_length);
+   if (!mic_st->final_frames)
+      goto error;
+
+   if (!mic_st->driver || !mic_st->driver->init)
+      goto error;
+
+   mic_st->driver_context = mic_st->driver->init();
+   if (!mic_st->driver_context)
+      goto error;
+
+   if (!string_is_empty(settings->arrays.microphone_resampler))
+      strlcpy(mic_st->resampler_ident,
+            settings->arrays.microphone_resampler,
+            sizeof(mic_st->resampler_ident));
+   else
+      mic_st->resampler_ident[0] = '\0';
+
+   mic_st->resampler_quality     = microphone_driver_get_resampler_quality(settings);
+   mic_st->source_ratio_original = (double)settings->uints.audio_output_sample_rate / settings->uints.microphone_sample_rate;
```

Comment:
`settings->uints.audio_output_sample_rate` should ideally be replaced by a value set by the core.
For example melonDS expects microphone input at a sample rate of 44.1 kHz.
When this proposed `core_input_sample_rate` is equal to the `microphone_sample_rate` you could also skip initializing and using the resampler (unless it has an optimized path in case the ratio is 1.0, I didn't check).